server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java

```diff
@@ -511,42 +500,118 @@ private void handleJoinRequest(JoinRequest joinRequest, ActionListener<Void> joi
+                    new ActionListener<>() {
+                        @Override
+                        public void onResponse(Empty empty) {
+                            validateListener.onFailure(e);
+                        }
+
+                        @Override
+                        public void onFailure(Exception e2) {
+                            e.addSuppressed(e2);
+                            validateListener.onFailure(e);
+                        }
+                    });
             }
         });
     }
 
+    private void sendJoinValidate(DiscoveryNode discoveryNode, ClusterState clusterState, ActionListener<Empty> listener) {
+        transportService.sendRequest(
+            discoveryNode,
+            JoinHelper.JOIN_VALIDATE_ACTION_NAME,
+            new ValidateJoinRequest(clusterState),
+            TransportRequestOptions.of(null, TransportRequestOptions.Type.STATE),
+            new ActionListenerResponseHandler<>(
+                listener.delegateResponse((l, e) -> {
+                    logger.warn(
+                        () -> new ParameterizedMessage(
+                            "failed to validate incoming join request from node [{}]",
+                            discoveryNode),
+                        e);
+                    listener.onFailure(new IllegalStateException("failure when sending a validation request to node", e));
+                }),
+                i -> Empty.INSTANCE,
+                Names.GENERIC));
+    }
+
+    private void sendJoinPing(DiscoveryNode discoveryNode, TransportRequestOptions.Type channelType, ActionListener<Empty> listener) {
+        if (discoveryNode.getVersion().onOrAfter(Version.V_8_0_0)) {
+            transportService.sendRequest(
+                discoveryNode,
+                JoinHelper.JOIN_PING_ACTION_NAME,
+                TransportRequest.Empty.INSTANCE,
+                TransportRequestOptions.of(null, channelType),
+                new ActionListenerResponseHandler<>(
+                    listener.delegateResponse((l, e) -> {
+                        logger.warn(
+                            () -> new ParameterizedMessage(
+                                "failed to ping joining node [{}] on channel type [{}]",
+                                discoveryNode,
+                                channelType),
+                            e);
+                        listener.onFailure(new IllegalStateException("failure when sending a join ping request to node", e));
+                    }),
+                    i -> Empty.INSTANCE,
+                    Names.GENERIC));
+        } else {
+            listener.onResponse(Empty.INSTANCE);
+        }
+    }
+
     private void processJoinRequest(JoinRequest joinRequest, ActionListener<Void> joinListener) {
         final Optional<Join> optionalJoin = joinRequest.getOptionalJoin();
-        synchronized (mutex) {
-            updateMaxTermSeen(joinRequest.getTerm());
+        try {
+            synchronized (mutex) {
```

Comment:
It is clearly fine now, but with the added complexity, I wonder if we should assert that we are not on the transport thread here?